Add Fortio-Envoy optimization guide (#29)
intel/optimization-zone: software/envoy/README.md

# Envoy + Fortio Benchmarking: Spin Lock Overhead & Optimization Guide

## Overview

> Evaluates Envoy running as a TCP proxy in front of Fortio, which acts as the backend load generator. The benchmark focuses on proxy-path performance and behavior under load, measuring metrics such as QPS and latency. Both server-side and client-side components are used to generate traffic and collect results, with Envoy and Fortio running in Docker containers based on the images listed below:

Reviewer: As a first time reader, topology is unclear. Is it possible to provide a diagram to explain setup? 
It is not clear what runs as the server vs client. 
Is Fortio being used as both load generator as well as server?
Is client on a different host?

> Evaluates Envoy running as a TCP proxy in front of Fortio, which acts as the backend load generator. The benchmark focuses on proxy-path performance and behavior under load, measuring metrics such as QPS and latency. Both server-side and client-side components are used to generate traffic and collect results, with Envoy and Fortio running in Docker containers based on the images listed below:

Reviewer: Consider starting with an overview followed by paragraphs explaining the two components, and then topology/setup used.  
### Overview 
This tuning guide describes best known practises to optimize performance....... when you run Fortio Envoy... 
### Fortio  
### Envoy  
### Topology/setup

Author: added diagram

```

---

## What Fortio Does

[Fortio](https://github.com/fortio/fortio) is a fast, multi-protocol load testing tool and echo server written in Go.

- **Server mode**: Listens on a port and echoes HTTP requests back. Minimal business logic - it is purely a throughput target.
- **Client mode**: Sends a configurable number of concurrent connections at a target QPS, collecting per-request latency histograms (p50, p90, p99, p99.9).
- **Why it matters here**: Fortio saturates Envoy so we can observe how the proxy performs under real concurrency.

---

## What Envoy Does

[Envoy](https://www.envoyproxy.io/) is a high-performance, C++ L4/L7 proxy used as the data plane in service meshes.

- **Event-driven, non-blocking I/O**: Each worker thread runs an independent libevent loop.
- **`--concurrency N`**: Spawns N worker threads. Each thread owns its own listener socket and connection pool, so there is near-zero cross-thread coordination for established connections.
- **TCP proxy mode** (used here): Envoy accepts a TCP connection on port 9090, opens a connection to Fortio on 8080, and shuttles bytes between them. No L7 parsing overhead.

> - **In mesh mode (`SECURE_MESH=true`)**: Adds mTLS - Envoy terminates the downstream TLS connection and re-originates a new TLS connection upstream, roughly doubling the cryptographic work per connection.

Reviewer: Earlier in the document you mentioned SECURE_MESH=true as “no Envoy sidecars, raw application performance” (direct mode), but here you have SECURE_MESH=true Envoy terminates the TLS connection?

Author: removed this comment.

---

## CPU Utilization and CPU Quota

> The script applies Docker CPU quotas (`--cpus 16` for Fortio, `--cpus 8` for Envoy). On a high core-count server (eg., 128 cores/256Threads), Docker enforces these quotas via cgroup CPU BW control. The OS spreads threads across all cores but throttles aggregate CPU time, resulting in roughly **6 - 7% per-core utilization** across all server cores - not saturation. The CPU quota is the binding constraint, not the WL.

Reviewer: Which script? You can provide an example script similar to this - https://github.com/intel/optimization-zone/blob/main/software/kafka/README.md#example-system-startup-script

---

## Spin Lock Overhead on High Core Count Machines

### What happens

On a high core-count server, running Fortio without concurrency limits causes significant `native_queued_spin_lock_slowpath` overhead. This behavior is observed with the fortio/fortio:1.71.1 image, which does not confine execution to a small number of CPU cores. Newer Fortio releases built with updated Go versions generally show reduced spin-lock overhead, as they tend to use fewer cores by default, hence better OOB QPS/latencies:

1. **Go runtime (Fortio)**: By default, Go sets `GOMAXPROCS` to the number of logical CPUs visible to the process. On a 128-core/256Threads machine, Go spawns up to 128 OS threads. The Go scheduler uses spin loops - a thread that finds its run queue empty will busy-spin for a short window before parking. With many threads occasionally spinning, aggregate spin overhead becomes significant.

2. **Cache coherency traffic**: Spin locks and atomic CAS operations on shared scheduler state cause cache line bouncing across all sockets. On a multi-socket NUMA system, cross-socket coherency traffic adds latency to every lock acquisition and scales with core count.

> 3. **Kernel paths involved** (from perf flame graphs):

Reviewer: Is it possible to upload a flamegraph for reference?

Author: added perf report

   - Futex contention: `runtime.lock()` -> `futex()` -> `native_queued_spin_lock_slowpath`
   - Go GC work-stealing: `gcAssistAlloc`, `gcDrainN`, `lfstack.pop`
   - TCP send/recv paths: `tcp_sendmsg`, `tcp_recvmsg`
   - Netpoll: `runtime.netpoll`, `netpollblock`

4. **Envoy `--concurrency` and NUMA**: If Envoy threads are allowed to migrate across NUMA nodes, each worker incurs NUMA-remote memory accesses and LLC  thrashing.

### Symptom

Latency p99/p99.9 climbs, throughput plateaus below the theoretical limit, and `perf` shows high `native_queued_spin_lock_slowpath`, `context-switches`. Spin lock overhead in perf traces is markedly higher in secure-mesh mode than proxy mode due to the additional TLS workload.

---

## Optimizations

### 1. NUMA Pinning (Most Impactful)

> Pin both Fortio and Envoy to a single NUMA node. This is the single most impactful optimization - it substantially reduces `native_queued_spin_lock_slowpath` overhead by keeping all memory allocations, thread migrations, and NIC interrupts on the same socket.

Reviewer: Pin both Fortio and Envoy to a single NUMA node. We mean on the server host?

```

---

## What Fortio Does

[Fortio](https://github.com/fortio/fortio) is a fast, multi-protocol load testing tool and echo server written in Go.

- **Server mode**: Listens on a port and echoes HTTP requests back. Minimal business logic - it is purely a throughput target.
- **Client mode**: Sends a configurable number of concurrent connections at a target QPS, collecting per-request latency histograms (p50, p90, p99, p99.9).
- **Why it matters here**: Fortio saturates Envoy so we can observe how the proxy performs under real concurrency.

---

## What Envoy Does

[Envoy](https://www.envoyproxy.io/) is a high-performance, C++ L4/L7 proxy used as the data plane in service meshes.

- **Event-driven, non-blocking I/O**: Each worker thread runs an independent libevent loop.
- **`--concurrency N`**: Spawns N worker threads. Each thread owns its own listener socket and connection pool, so there is near-zero cross-thread coordination for established connections.
- **TCP proxy mode** (used here): Envoy accepts a TCP connection on port 9090, opens a connection to Fortio on 8080, and shuttles bytes between them. No L7 parsing overhead.

---

> ## Topology / Setup

Reviewer: Thanks for adding this! Looks good.

```mermaid

flowchart RL
    subgraph CLIENT[Client Machine - network host B]
        FC["Fortio Client\ncpus 16, GOMAXPROCS=16\nconcurrency=N, qps=0, t=30s"]
        CE["Envoy client-side\ncpus 8, concurrency 8\nlisten :9091"]
    end

    subgraph SERVER[Server Machine - network host A]
        SE["Envoy server-side\ncpus 8, concurrency 8\nlisten :9090"]
        FS["Fortio Server\ncpus 16, GOMAXPROCS=16\nlisten :8080"]
    end

    FC -- "1. localhost:9091" --> CE
    CE -- "1. TCP to REMOTE_IP:9090" --> SE
    SE -- "1. forward to 127.0.0.1:8080" --> FS
    FC -. "2. direct-bench to REMOTE_IP:8080" .-> FS
```

> **1.** Default proxy mode - traffic traverses both Envoy proxies.  
> **2.** `direct-bench(secure mesh)` mode - both Envoys (side cars) are bypassed. Fortio client hits Fortio server directly.  

> > **3.** Server and Client are run on different machines & Clinet side measures QPS/latencies.  

Reviewer: Client

1. Writes `envoy_client.yaml` pointing to `REMOTE_IP:9090`
2. Starts **Envoy client-side** listening on `localhost:9091`
3. Runs **Fortio client** sending load to `localhost:9091`

**direct-bench mode** — skips both Envoys and sends Fortio traffic directly to `REMOTE_IP:8080`. Optionally sets secure mesh labels (`UBER_NET_CLASSID`, `com.uber.secure_service_mesh`) if `SECURE_MESH=true`.

Key configs:

| Parameter | Fortio client | Envoy client-side |
|---|---|---|
| `--cpus` | 16 | 8 |
| `GOMAXPROCS` | optional, via env | N/A |
| `-c` (concurrency) | `CONCURRENCY` env (default 2000) | N/A |
| `-t` (duration) | 30s | N/A |
| `--payload-size` | 5000 bytes | N/A |
| `-qps` | 0 (max throughput) | N/A |
| listen port | N/A | 9091 |
| `--concurrency` | N/A | 8 |
| circuit breaker | N/A | max_connections=20000 |

```bash
# Default mode: Fortio -> client Envoy :9091 -> server Envoy :9090 -> Fortio server :8080
GOMAXPROCS=16 CONCURRENCY=1000 ./client.sh <server-ip>

# Direct-bench mode: Fortio -> Fortio server :8080 (no Envoy)
GOMAXPROCS=16 SECURE_MESH=true CONCURRENCY=1000 ./client.sh <server-ip> direct-bench
```

---

## CPU Utilization and CPU Quota

> The script applies Docker CPU quotas (`--cpus 16` for Fortio, `--cpus 8` for Envoy). On a high core-count server (eg., 128 cores/256Threads), Docker enforces these quotas via cgroup CPU BW control. The OS spreads threads across all cores but throttles aggregate CPU time, resulting in roughly **6 - 7% per-core utilization** across all server cores - not saturation. The CPU quota is the binding constraint, not the WL.

Reviewer: WL > Workload

2. **Cache coherency traffic**: Spin locks and atomic CAS operations on shared scheduler state cause cache line bouncing across all sockets. On a multi-socket NUMA system, cross-socket coherency traffic adds latency to every lock acquisition and scales with core count.

3. **Kernel paths involved** (from `perf report` — see [Perf Report](#perf-report-baseline) below):
   - Futex contention: `runtime.lock()` -> `futex()` -> `native_queued_spin_lock_slowpath`
   - Go GC work-stealing: `gcAssistAlloc`, `gcDrainN`, `lfstack.pop`
   - TCP send/recv paths: `tcp_sendmsg`, `tcp_recvmsg`
   - Netpoll: `runtime.netpoll`, `netpollblock`

4. **Envoy `--concurrency` and NUMA**: If Envoy threads are allowed to migrate across NUMA nodes, each worker incurs NUMA-remote memory accesses and LLC  thrashing.

### Symptom

Latency p99/p99.9 climbs, throughput plateaus below the theoretical limit, and `perf` shows high `native_queued_spin_lock_slowpath`, `context-switches`. Spin lock overhead in perf traces is markedly higher in secure-mesh mode than proxy mode due to the additional TLS workload.

---

## Perf Report (Baseline) <a name="perf-report-baseline"></a>

> `perf report` output from a baseline run (no optimizations applied), 112K samples of `cycles:P`:

Reviewer: Thanks for adding this.

- **`gcDrain` (2.56%) + `gcDrainN` (1.14%) + GC atomics** — GC scanning overhead; high allocation rate from concurrent Fortio requests forces frequent GC cycles
- **`procyield` (1.04%) + `lock2` (0.86%) + `stealWork` (0.67%)** — Go scheduler spin-waiting and work-stealing across processors, worsened by high `GOMAXPROCS`

---

## Optimizations

### 1. NUMA Pinning (Most Impactful)

Pin both Fortio and Envoy to a single NUMA node on server side. This is the single most impactful optimization - it substantially reduces `native_queued_spin_lock_slowpath` overhead by keeping all memory allocations, thread migrations, and NIC interrupts on the same socket.

```bash
# Pin both containers to NUMA node 0
sudo docker run ... --cpuset-cpus "<numa0 CPU range based on the SKU>" --cpuset-mems "0" fortio/fortio:1.71.1 ...
sudo docker run ... --cpuset-cpus "<numa0 CPU range based on the SKU>" --cpuset-mems "0" envoyproxy/envoy:v1.31.10 ...
```

Or on the host directly:

```bash
numactl --cpunodebind=0 --membind=0 -- envoy -c envoy.yaml --concurrency 16
```

> **Why it helps**: Cross-socket coherency traffic is the dominant cause of `native_queued_spin_lock_slowpath` overhead on high core-count systems. Confining the WL to NUMA node 0 eliminates this entirely.

Reviewer: WL > Workload

---

### 2. Increase CPU Quota for Both Containers

> With NUMA pinning in place, increasing the CPU quota for both Fortio and Envoy containers further reduces scheduling delays and spin lock overhead. Keep `--concurrency` equal to `--cpus` to avoid over-subscription. The numbers are just examples. Tune this based on the SKU/cores used. 

Reviewer: On 256 threads system, you have allocated only 32+16 cores? What would be a good general recommendation? If someone had to tune based on SKU/cores they have?

Author: There are no formulas that can be used.. Just general guidance is what can be provided

---

### 4. GC Overhead Reduction - Go's GreenTea GC

Fortio's load generator creates large numbers of short-lived objects (request/response structs, buffers, timers). The default Go GC (tricolor mark-and-sweep) scans the object graph one object at a time, causing poor spatial locality, high contention on global queues, and significant cycles spent in the scan loop.

**GreenTea GC** (prototype in Go 1.24, available in Go 1.25.1(as exp feature)) is a span-centric generational collector. Should be enabled by default in Go 1.26:

- Scans in aligned 8 KB spans rather than individual objects -> better cache behavior and less evictions.
- Reduces overhead from `gcDrain`, `trygetfull`, and `lfstack` paths observed in perf traces.
- New span operations (`tryDeferToSpanScan`, `localSpanQueue.stealFrom`) distribute GC work more evenly across processors.
- Results in less time spent in GC and more CPU available for the application.

To use GreenTea GC, build Fortio with Go 1.25.1 and the GreenTea flag enabled or use 1.26 where its enabled by default.

---
### 5. Use fewer CPU cores as possible to run Envoy + fortio

> Fortio and Envoy are intentionally run on a reduced number of CPU cores to avoid excessive spin-lock contention and busy-waiting behavior. By limiting core availability, the benchmark prevents threads from continuously spinning on shared locks and instead exposes meaningful contention, scheduling, and proxy-path behavior under realistic CPU pressure 

Reviewer: Is the suggestion to increase docker cpu quota but restrict to fewer CPU cores? This is a bit confusing. Can you clarify/elaborate? 
should cpu quota match the number of cores it is being pinned to 
i.e. --cpus = -cpuset-cpus "<numa0 CPU range based on the SKU>"

Author: cpus and cpuset-cpus should be same.

---
### 6. Other Envoy Tuning -- which can be explored

#### Worker and socket tuning

> - **`SO_REUSEPORT`**: Already used by Envoy by default. Verify it is not disabled by any sysctls - it distributes `accept()` load evenly across worker threads without a shared accept mutex.

Reviewer: Will be useful to provide which sysctl setting to verify

> - **`--concurrency`**: Tune thread concurrency to match the number of physical cores assigned to the container. Over-provisioning causes false sharing, under-provisioning wastes hardware.

Reviewer: Same as bullet 2? --concurrency equal to --cpus

#### Huge pages (TLB pressure)
Envoy's memory allocator (`tcmalloc` / `jemalloc`) benefits from 2 MB huge pages, which reduce TLB miss rates when proxying many concurrent flows:

```bash
echo 512 > /proc/sys/vm/nr_hugepages
```

#### CPU frequency governor & BIOS/OS settings
Disable dynamic frequency scaling to eliminate governor-induced latency spikes:

```bash
cpupower frequency-set -g performance
```

> Ensure you have configured the floowing as deflaut in BIOS or from the OS(using perfspect tool) on Granite Rapids (Xeon6) or later systems 

Reviewer: Following

1. `Efficiency Latency Control`: Latency Optimized
2. `Energy Performance Bias`: Performance (0)
3. `Energy Performance Preference`:  Performance (0)

---

## Quick Diagnosis Checklist

| Symptom | Likely cause | Fix |
|---|---|---|
| High `native_queued_spin_lock_slowpath` in perf | Cross-NUMA memory access | Pin both containers to NUMA node 0 |
| High TLB in perf or spinlocks | Cross-NUMA memory access | Ensure Huge Pages are enabled |
| High `native_queued_spin_lock_slowpath` from Go threads | Too many Go OS threads | Set `GOMAXPROCS` = `--cpus` value |
| High LLC-load-misses in `perf stat` | Cross-NUMA memory access | Pin to single NUMA node |
| GC overhead in perf traces (`gcDrain`, `trygetfull`) | High allocation rate with default GC | Build Fortio with GreenTea GC (Go 1.25.1); raise `GOGC` |

> | Envoy CPU bottlenecked in TLS | mTLS handshake overhead | Enable TLS session resumption. Make TLS communicaiton/handshake Async |

Reviewer: communication

> | NIC softirq on same cores as Envoy | IRQ affinity not set or not part of of same NUMA | Separate NIC IRQ cores from Envoy worker cores or ensure cores + mem + NIC are on same NUMA |

Reviewer: duplicate of

Author: did not understand this comment

Reviewer: ```suggestion
| NIC softirq on same cores as Envoy | IRQ affinity not set or not part of same NUMA | Separate NIC IRQ cores from Envoy worker cores or ensure cores + mem + NIC are on same NUMA |
```

> **3.** Server and Client are run on different machines & Client side measures QPS/latencies.  
> **4.** Setup: Used a high core count machine such as Intel Xeon 6 (Granite Rapids)-- 6980P (128C/256T) for both server & Client

---

## Scripts Overview

### server.sh

Starts two Docker containers on the server machine using host networking:

1. **Fortio server** — echo HTTP server, accepts traffic on `:8080`
2. **Envoy server-side** — TCP proxy, listens on `:9090` and forwards to `127.0.0.1:8080`

Key configs:

| Parameter | Fortio | Envoy |
|---|---|---|
| `--cpus` | 16 | 8 |
| `GOMAXPROCS` | optional, via env | N/A |
| listen port | 8080 | 9090 |
| `--concurrency` | N/A | 8 |
| circuit breaker | N/A | max_connections=20000 |

```bash
# Run on the server machine
./server.sh
```

---

### client.sh

Starts containers on the client machine and drives load toward the server. Supports two modes:

**Default Proxy mode** — spins up a client-side Envoy that proxies to the server, then runs Fortio through it:

1. Writes `envoy_client.yaml` pointing to `REMOTE_IP:9090`
2. Starts **Envoy client-side** listening on `localhost:9091`
3. Runs **Fortio client** sending load to `localhost:9091`

> **direct-bench mode** — skips both Envoys and sends Fortio traffic directly to `REMOTE_IP:8080`. Optionally sets secure mesh labels (`UBER_NET_CLASSID`, `com.uber.secure_service_mesh`) if `SECURE_MESH=true`.

Reviewer 2: @vaibhavk2 Are these secure mesh labels specific to the server.sh and client.sh used for your testing? Are we planning on providing those scripts?

Author: These are the same scripts. These are customer shared scripts (by Uber). I don't want to share those. But i have highlighted all the requirements/spec of the script, so people can create those if needed.

Reviewer 2: Okay in light of that I think we can remove the secure mesh labels comment then. Instead, just:

```suggestion
**direct-bench mode** — skips both Envoys and sends Fortio traffic directly to `REMOTE_IP:8080`. 
```

**eg — 256 vCPU dual-socket BM (64 physical cores per NUMA node):**
- Choose Envoy workers = 16 -> Fortio GOMAXPROCS = 32 -> total = 48 &le; 64
- `--cpus 16 --cpuset-cpus "0-15" --cpuset-mems 0 --concurrency 16` for Envoy
- `--cpus 32 --cpuset-cpus "16-47" --cpuset-mems 0 -e GOMAXPROCS=32` for Fortio

**eg — 16 vCPU VM (8 physical cores, 1 NUMA node):**
- Choose Envoy workers = 2 -> Fortio GOMAXPROCS = 4 -> total = 6 ≤ 8
- `--cpus 2 --cpuset-cpus "0-1" --cpuset-mems 0 --concurrency 2` for Envoy
- `--cpus 4 --cpuset-cpus "2-5" --cpuset-mems 0 -e GOMAXPROCS=4` for Fortio

Three settings must be consistent on each container, or spin locks return:

- `--cpuset-cpus` and `--cpus` must cover the **same** cores — `--cpus` is a quota and `--cpuset-cpus` is the hard pin. They must match
- `--cpuset-mems` must be set to the **same NUMA node id** — without it the kernel allocates memory on a remote socket even when CPUs are pinned
- Envoy `--concurrency` must equal its `--cpus` value, and Fortio `GOMAXPROCS` must equal its `--cpus` value — extra threads beyond available cores spin on empty queues causing `native_queued_spin_lock_slowpath`

> ### Reference — Envoy workers and NUMA fit check across SKUs

Reviewer: Looks awesome! Thank you for putting this together.

# Envoy + Fortio Benchmarking: Spin Lock Overhead & Optimization Guide

> 

Reviewer 2: Since there are a lot of sections here, it would be nice to have a Table of Contents to help readers quickly identify where they might want to start.

## Overview

> This tuning guide describes best known practices to optimize Fortio + Envoy performance. It evaluates Envoy running as a TCP proxy in front of Fortio, which acts as the backend load generator. The benchmark focuses on proxy-path performance and behavior under load, measuring metrics such as QPS and latency. Both server-side and client-side components are used to generate traffic and collect results, with Envoy and Fortio running in Docker containers based on the images listed below:

Reviewer 2: I would suggest removing something like the following:

```suggestion
This tuning guide describes best known practices to optimize Envoy TCP proxy performance. Fortio is used as a load generator, which helps to reveal bottlenecks and test optimizations. This benchmark configuration focuses on proxy-path performance and behavior under load, measuring metrics such as QPS and latency.
```

I would put this part in a later section: "Both server-side and client-side components are used to generate traffic and collect results, with Envoy and Fortio running in Docker containers based on the images listed below:

- **Fortio**: `fortio/fortio:1.71.1`
- **Envoy**: `envoyproxy/envoy:v1.31.10`"

- **TCP proxy mode** (used here): Envoy accepts a TCP connection on port 9090, opens a connection to Fortio on 8080, and shuttles bytes between them. No L7 parsing overhead.

---

## Topology / Setup

```mermaid

flowchart RL
    subgraph CLIENT[Client Machine - network host B]
        FC["Fortio Client\ncpus 16, GOMAXPROCS=16\nconcurrency=N, qps=0, t=30s"]
        CE["Envoy client-side\ncpus 8, concurrency 8\nlisten :9091"]
    end

    subgraph SERVER[Server Machine - network host A]
        SE["Envoy server-side\ncpus 8, concurrency 8\nlisten :9090"]
        FS["Fortio Server\ncpus 16, GOMAXPROCS=16\nlisten :8080"]
    end

    FC -- "1. localhost:9091" --> CE
    CE -- "1. TCP to REMOTE_IP:9090" --> SE
    SE -- "1. forward to 127.0.0.1:8080" --> FS
    FC -. "2. direct-bench to REMOTE_IP:8080" .-> FS
```

> **1.** Default proxy mode - traffic traverses both Envoy proxies.  
> **2.** `direct-bench(secure mesh)` mode - both Envoys (side cars) are bypassed. Fortio client hits Fortio server directly.  
> **3.** Server and Client are run on different machines & Client side measures QPS/latencies.  

> > **4.** Setup: Used a high core count machine such as Intel Xeon 6 (Granite Rapids)-- 6980P (128C/256T) for both server & Client

Reviewer 2: ```suggestion
> **4.** Setup: The bottlenecks identified in this analysis tend to happen on Client and Server machines with large core counts (e.g. >= 96)
```

> 

Reviewer 2: After the topology/setup is where I think it makes sense to capture the Fortio/Envoy versions.

---

> ## Scripts Overview

Reviewer 2: Since we're not providing any scripts, I think it makes more sense to call this Client/Server Configuration.
```suggestion
## Client/Server Configuration
```

### server.sh

Starts two Docker containers on the server machine using host networking:

1. **Fortio server** — echo HTTP server, accepts traffic on `:8080`
2. **Envoy server-side** — TCP proxy, listens on `:9090` and forwards to `127.0.0.1:8080`

Key configs:

| Parameter | Fortio | Envoy |
|---|---|---|
| `--cpus` | 16 | 8 |
| `GOMAXPROCS` | optional, via env | N/A |
| listen port | 8080 | 9090 |
| `--concurrency` | N/A | 8 |
| circuit breaker | N/A | max_connections=20000 |

```bash
# Run on the server machine
./server.sh

> ```

Reviewer 2: I think it makes sense to remove this

1. Writes `envoy_client.yaml` pointing to `REMOTE_IP:9090`
2. Starts **Envoy client-side** listening on `localhost:9091`
3. Runs **Fortio client** sending load to `localhost:9091`

**direct-bench mode** — skips both Envoys and sends Fortio traffic directly to `REMOTE_IP:8080`. Optionally sets secure mesh labels (`UBER_NET_CLASSID`, `com.uber.secure_service_mesh`) if `SECURE_MESH=true`.

Key configs:

| Parameter | Fortio client | Envoy client-side |
|---|---|---|
| `--cpus` | 16 | 8 |
| `GOMAXPROCS` | optional, via env | N/A |
| `-c` (concurrency) | `CONCURRENCY` env (default 2000) | N/A |
| `-t` (duration) | 30s | N/A |
| `--payload-size` | 5000 bytes | N/A |
| `-qps` | 0 (max throughput) | N/A |
| listen port | N/A | 9091 |
| `--concurrency` | N/A | 8 |
| circuit breaker | N/A | max_connections=20000 |

```bash
# Default mode: Fortio -> client Envoy :9091 -> server Envoy :9090 -> Fortio server :8080
GOMAXPROCS=16 CONCURRENCY=1000 ./client.sh <server-ip>

# Direct-bench mode: Fortio -> Fortio server :8080 (no Envoy)
GOMAXPROCS=16 SECURE_MESH=true CONCURRENCY=1000 ./client.sh <server-ip> direct-bench

> ```

Reviewer 2: I think it makes sense to remove this block since we're not providing the scripts
